power: Add device idle power management support

ramakrishnapallala opened the pull request:
Add framework for device Idle Power Management(IPM)
for suspending devices based on device idle. This will
help in saving power even while system(CPU) is active.

This framework uses API's like device_set_power_state
and device_get_power_state to get/set the present
device power state. It relies on device_busy_check API
to know whether the device is idle or busy before
taking suspend or resume action.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

pizi-nordic commented on include/device.h:
```diff
@@ -439,6 +443,70 @@ int device_any_busy_check(void);
+ * @retval 0 If successful.
+ * @retval Errno Negative errno code if failure.
+ */
+int device_ipm_get(struct device *dev);
```
General comments:
- The asynchronous interface needs a callback to notify when device is ready to be used or when it entered the idle mode. It also has to support transitions triggered in the middle of other transitions (one thread called put() and other get() almost simultaneously but device needs XXXms to do transition).
- The asynchronous calls needs to trace multiple users due to separate callbacks (there could be several subsystems requesting to either enable or disable power saving).

ramakrishnapallala replied:
we can add notification mechanism for async calls and pass the error no to let the driver know the status.
what do you mean by "trace multiple users"?

pizi-nordic replied:
Please consider device which needs some time to enter sleep mode or wake-up (I2C-connected gyroscope for example). The gyro is used by 2 mutually exclusive functionalities. What will happen if one functionality will be enabled just after disabling another?

Func1: Disabled
-> Gyro is now idle, we can go to suspend state.
-> Driver sends SUSPEND command (to be more specific schedules an I2C transaction).
-> Now driver waits for I2C transaction result, so other things can be scheduled.

Func2: Enabled
-> What is state of the Gyro? Should we send Wakeup command?

If Func1 and Func2 are using blocking calls, the lock on the device will solve this problem, however the Func1 will sleep with lock held . If one of this functionality would like to use asynchronous API, we have to trace the state of the device in the PM infrastructure.

pizi-nordic commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,182 @@
+		goto out;
+	}
+
+	ret = device_set_power_state(dev, cfg->ipm_request);
```
If change of device power state is long, you are blocking while tread executing delayed works in the system.
The same applies for k_sem_take() above.

ramakrishnapallala replied:
I think we should change the semaphore to spinlock as it would be the right primitive to use here.

On device_set_power_state call, we will have to execute this call at some point, even if it takes longer. How can we address this? eventually, we will have to call this right?

pizi-nordic replied:
I have worked with gyro which needed almost 100ms for wake-up. We have to design the interlace to deal with such devices. IMHO the device_set_power_state() should be asynchronous. Otherwise we will end locking something for 100ms. It will be workqueue, calling thread or whole system, depending where we put the lock.

IMHO Device PM needs an FSM and trace all the transitions. I would like to propose the following scheme:

1. Variables:
- `curent_state`
- `in_transition`
- `target_state`

2. API:
```
state_changed_cb(state)
{
 current_state = state;
 if (target_state != current_state)
    trigger_state_change(target_state, state_changed_cb)
else
  in_transition = false
}
 
set_powert_state(state)
{
  target_state = state
  if (!in_transition && (target_state != current_state)) {
    trigger_state_change(target_state, state_changed_cb)
    in_transition = true;
  } 
}
```

pizi-nordic replied:
@ramakrishnapallala: I will hold with review until we get an idea how to support devices which needs a lot of time to change the state.

ramakrishnapallala replied:
@pizi-nordic even if we make device_set_power_state() asynchronous the problem will remain the same. Th main issue would the time the device takes to suspend/resume. Until those suspend/resume calls are executed the API's which use these devices will be blocked. In case asynchronous pm_get/put we can add the timeout for the delayed work which would be helpful in case back to back transitions.
If we think the time taken to resume or suspend is not worth power saving or impacts the expected performance then the driver can simply not implement the device idle PM.

pabigot replied:
I've gotta make my usual complaint: what exactly are these functions supposed to do?  Is there design documentation somewhere?  The header only specifies that functions enable, disable, suspend, and resume, with no information about exactly what that's supposed to mean, especially if these operations may become asynchronous or multiple devices are being turned down in sequence.  How do they interact?  What are the possible error codes and what specifically do they mean in this context?

I'm not saying this shouldn't be merged--I haven't been able to look at it closely enough--but I am saying that I have no faith that this API will do what I want when I get around to using LPM (which is very important to me).

ramakrishnapallala replied:
@pabigot I will update the PR with more documentation. Thanks.

pizi-nordic replied:
> The only thing is it uses the system thread to execute the handler.

And this is my concern. Switching power state of external device connected to SPI or I2C might take hundreds of us. During that time system thread servicing work queue will be blocked. We can mitigate this by allowing registration of custom thread for PM management. What do you think?

ramakrishnapallala replied:
yeah, that makes a lot of sense. I will add the support and update the PR. Thanks.

pabigot replied:
@pizi-nordic can you point me to an external device that has the behavior you describe?  I'd like to understand the operations involved and the cause of the delay to see if an approach I'm prototyping that gets rid of using threads within sensor drivers might apply.

pizi-nordic replied:
@pabigot: Here you have few examples:
- Any device connected to I2C: For bus speed 100kHz, 8-bit device address, 8-bit register address and 8-bit value written there to put device in idle mode, you will spent at least 270us waiting for I2C transaction to complete. 

- LIS3DH Accelerometer: Change of operating mode takes 1/ODR (ODR - internal clock called "Data Rate" in the datasheet, which could be as low as 1Hz) + 1.6ms. 

- Any device with analog part which needs some time to reach operating point (PLL, Accelerometer, Gyroscope, Temp Sensor and so on).

pabigot replied:
Thanks; that's the sort of thing I was expecting.

pizi-nordic commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,182 @@
+	int ret;
+	u32_t power_state;
+
+	atomic_dec(&dev->config->usage);
```
You are decreasing usage count, but suspend is continued regardless of the value of usage counter.

ramakrishnapallala replied:
we will check the existing state and then decide whether to call the set_power_state or not. we can add some checks to limit the count value.

pizi-nordic commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,182 @@
+	}
+
+	/* Return error if device is busy */
+	if (device_busy_check(dev)) {
```
If usage is 0 and device is busy we have a serious problem in the power management infrastructure. This should be a critical bug.

ramakrishnapallala replied:
Yes, this is more like safety check. Like, the driver can set the device busy state and block the suspend.

pizi-nordic commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,182 @@
+	}
+
+	/* Cancel already queued work */
+	k_delayed_work_cancel(&dev->config->ipm_work);
```
I do not know if we can do that without  tracing device state. The driver will see multiple calls to device_set_power_state with the same argument.

ramakrishnapallala replied:
Ok will rework the logic here.

pizi-nordic commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,182 @@
+
+}
+
+static int device_ipm_resume(struct device *dev, u32_t ipm_flags)
```
Please see comments for device_ipm_suspend()

pizi-nordic commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,182 @@
+
+}
+
+int device_ipm_get(struct device *dev)
```
IMHO usage could should be traced regardless of ipm_enable, as we might enable/disable power management when device is already used.,

ramakrishnapallala replied:
OK

pizi-nordic reviewed (changes requested):
IMHO we have to take under consideration that power state transition could take a lot of time (hundreds of ms), especially for sensors and there might be conflicting requests to enable and disable device during state transition. We should handle that without blocking the system. I think that this might meed making device_pm_control() an asynchronous call.

pizi-nordic commented:
BTW: Could we use "pm" instead of "ipm" in the function names?

ramakrishnapallala commented on include/device.h:
```diff
@@ -222,6 +221,11 @@ struct device_config {
 #ifdef CONFIG_DEVICE_POWER_MANAGEMENT
```
OK

pabigot replied:
There already is a `device_pm_control` structure that does something a little different; can this functionality be merged with that?

ramakrishnapallala commented on include/device.h:
```diff
@@ -439,6 +443,70 @@ int device_any_busy_check(void);
+ * @param dev Pointer to device structure of the specific device driver
+ * the caller is interested in.
+ */
+void device_ipm_init(struct device *dev);
```
Ok will move the logic accordingly.

ramakrishnapallala commented:
@pizi-nordic please have a look at the updated PR.

pabigot commented on include/device.h:
```diff
@@ -208,6 +208,24 @@ extern "C" {
+ */
+struct device_pm {
+	atomic_t usage;
+	atomic_t enable;
```
A question on usage of `atomic` in Zephyr:  Of these four values the only one that appears to be used in a read-modify-write context is `usage`.  All others get and set single-word complete values (at least, based on a quick scan of the diff).  Is there a need for access to them to be atomic?

If yes, I'm concerned that a lot of other state values in Zephyr would have the same problem and there's major cleanup to do.

If no, use of atomic (e.g. with `enable`) is confusing, suggesting there's something tricky going on, and better to change them to int, bool, or enum values.

ramakrishnapallala replied:
My intention is to avoid the use of lock here and want to make sure we don't run into the case where one thread write and other one reads at the same time. what would you suggest?

carlescufi replied:
@andrewboie @andyross can you comment on the atomic side of things?

pizi-nordic replied:
The usage of atomics in the RMW contexts is obvious. Using atomics in non-RMW contexts is also valid in my opinion if saves us from using classic lock. Please look in the example below: 

```
struct foo {
    atomic_t enabled;
    (...)
}

void foo_enable(struct foo *foo)
{
  if (atomic_set(&foo->enabled, 1) == 0)
  {
      /* Protected code */
  }
  (...)
}

void foo_disable(struct foo *foo)
{
  if (atomic_set(&foo->enabled, 0) == 1)
  {
      /* Protected code */
  }
  (...)
}
```

The code above ensured that the `/* Protected code */` will be executed only once even if multiple threads would call simultaneously either `foo_enable()` or `foo_disable()`. In such situation we avoided classic lock with its penalties.

However such protection will fail if call to `foo_enable()` could be pre-empted by `foo_disable()` (or in reverse). In such situation usage if classic lock is unavoidable.

I do not have access to the discussed code (it was updated), but in general I would recommend using classic locks for non-RMW scenarios as a safer choice.

ramakrishnapallala commented:
@pizi-nordic @pabigot 

Updated the PR, please have a look.

Changelog:
- Added worker thread for PM requests.
- Removed all atomic_t variables except usage and added a semaphore for synchronization.
- Merged device_pm_suspend/resume internal functions.
- Add more comments for better readability.

pabigot commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,238 @@
+}
+
+int device_pm_put_sync(struct device *dev)
+{
```
Both synchronous and asynchronous variants of `pm_put` (`pm_get`) are identical except for the flags passed to the underlying `set_state`.  Put the put (get)  implementation in a static function that takes the flags as a parameter and call it from the sync and async variants with the appropriate argument, so we don't have to worry about the locking logic being right in one case but not the other.  You might even be able to pass the state as an argument and combine all four into one, though that might get confusing.

ramakrishnapallala replied:
Ok thanks.

ramakrishnapallala replied:
this is fixed.

pabigot reviewed (dismissed):
I'm pretty sure `usage` is not being used correctly; see comments.

ramakrishnapallala commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,238 @@
+{
+	int ret;
+
+	if (atomic_inc(&dev->pm->usage) >= UINT_MAX) {
```
I think we remove this increment check, in other parts of the Zephyr code also there is not check for the atomic increment.

ramakrishnapallala commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,238 @@
+{
+	int ret;
+
+	if (atomic_dec(&dev->pm->usage) <= 0) {
```
the suspend should be invoked when the count is 0.

If the count is 0 which means no one is using the device so we can suspend it to save power.

pizi-nordic commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,226 @@
+	return device_pm_handle_put(dev, 0);
+}
+
+void device_pm_enable(struct device *dev)
```
IMHO there are 2 use-cases missed:
- Device with PM disabled, not used: After `device_pm_enable()` should be suspended.
- Device with PM enabled, not used: After `device_pm_disable()` should be woken up.

ramakrishnapallala replied:
Ok got it. Thanks.

pizi-nordic commented on include/device.h:
```diff
@@ -150,11 +150,18 @@ extern "C" {
 	};								  \
+	static struct device_pm _CONCAT(__pm_, dev_name) __used           \
+							= {               \
+		.name = drv_name,					  \
```
Could we either replace name with pointer to device or embed `device_pm` in `device`?
IMHO this would reduce work needed to traverse from one structure to another.

ramakrishnapallala replied:
the issue is, the device data is being put in init section which is read-only.

 static struct device _CONCAT(__device_, dev_name) __used          \
        **__attribute__((__section__(".init_" #level STRINGIFY(prio)))) = { \**

Due to this i had to use pointer here.

pizi-nordic replied:
Ok. Are you going to replace the name by the pointer to device structure?

ramakrishnapallala replied:
yes will fix it.

pizi-nordic commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,226 @@
+	}
+
+	/* Return if current state is same as target state */
+	if (current_state == target_state) {
```
Possible bug. Scenario: 
Async request to suspend followed by sync to wake-up.
1. Asynchronous request is queue in work queue, but not yet executed.
2. Sync request exists as the device is in target state.
3. Asynchronous request is executed, putting device in suspend state while it is used.

ramakrishnapallala replied:
Ok, I will have to move the worker cancel logic above this block.

pizi-nordic commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,226 @@
+		goto get_out;
+	}
+
+	ret = device_pm_set_state(dev,
```
I think that we should request new pm state only if usage count crosses 0/1 boundary.

ramakrishnapallala replied:
Ok.

pizi-nordic replied:
I am not getting this. IMHO we have to track only 2 things: What state transition is executing and what is requested state. This gives us 4-state FSM:
- DEVICE_ACTIVE:
- DEVICE_SUSPENDING:
- DEVICE_WAKING_UP:
- DEVICE_SHUTTING_DOWN:

Depending on the request and state, you can easily make the decision what to do and exit form the sync request when given state was reached.

ramakrishnapallala replied:
Ok got it. I will add the FSM logic to distinguish the states.

ramakrishnapallala replied:
As we wanted to avoid longer wait on async calls, I introduced separate lock for handling the power state transitions and during this time, the worker will not hold the normal lock. Due to this we may not be able to implement the above logic. I will update the PR please have a look.

pizi-nordic commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,226 @@
+		goto put_out;
+	}
+
+	ret = device_pm_set_state(dev,
```
I think that we should request new pm state only if usage count crosses 0/1 boundary.

ramakrishnapallala replied:
Ok.

pizi-nordic commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,226 @@
+		k_delayed_work_submit_to_queue(&devpm_workq,
+					&dev->pm->dwork, dev->pm->timeout);
+	} else {
+		ret = device_set_power_state(dev, target_state);
```
In the `pm_work_handler` we are doing `device_busy_check()`. Here it is skipped. Why?

ramakrishnapallala replied:
This is a synchronous call so the assumption is nothing changes on the device side that's why it was skipped.

On second thought, I think it's better to add the check here as well.

ramakrishnapallala commented on include/device.h:
```diff
@@ -447,6 +481,93 @@ int device_any_busy_check(void);
+ * @brief Call device resume asynchronously based on usage count
+ *
+ * Called by a device driver to mark the device as being used.
+ * This API will schedule a worker to asynchronously bring the
```
the called need wait on the alert from the worker. But the API will not return any specific code to differentiate b/w ready and queued. do you think we need return code? the caller anyway has to use the alert wait

pizi-nordic replied:
I think that the information about current status is needed. I looked into the code to check if we can notify wrong state, but this seems to be not possible due to locking.

However I found other thing: Both `pm_work_handler()` and `device_pm_handle_*()` shares the same lock. Which mean that if device need a long time to perform transition, the asynchronous interface will suck on the lock waiting for exiting from `pm_work_handler()`. I think that blocking asynchronous interface might be not welcome.

ramakrishnapallala replied:
I will try to limit the locking scope.

ramakrishnapallala replied:
I have introduced lock to control the device pm state transitions.

ramakrishnapallala commented on include/device.h:
```diff
@@ -447,6 +481,93 @@ int device_any_busy_check(void);
+ * @brief Call device suspend asynchronously based on usage count
+ *
+ * Called by a device driver to mark the device as being free.
+ * This API will schedule a worker to asynchronously put the
```
same as above, but in case suspend the caller need not wait for the suspend operation to complete.

ramakrishnapallala reviewed (commented):
Thanks for the comments

ramakrishnapallala commented on samples/subsys/power/device_pm/src/dummy_driver.c:
```diff
@@ -0,0 +1,95 @@
+
+int dummy_init(struct device *dev)
+{
+	device_power_state = DEVICE_PM_SUSPEND_STATE;
```
OK

ramakrishnapallala commented on samples/subsys/power/device_pm/src/dummy_driver.c:
```diff
@@ -0,0 +1,95 @@
+static int dummy_suspend(struct device *dev)
+{
+	printk("suspending..\n");
+	device_power_state = DEVICE_PM_SUSPEND_STATE;
```
I will add this in my upcoming PR's.

ramakrishnapallala commented on samples/subsys/power/device_pm/src/dummy_driver.c:
```diff
@@ -0,0 +1,95 @@
+static int dummy_device_pm_ctrl(struct device *dev, u32_t ctrl_command,
+				void *context)
+{
+	if (ctrl_command == DEVICE_PM_SET_POWER_STATE) {
```
Ok.

ramakrishnapallala commented on samples/subsys/power/device_pm/src/main.c:
```diff
@@ -0,0 +1,25 @@
+	struct dummy_driver_api *api;
+	int ret, val;
+
+	printk("Device PM sample app start\n");
```
Ok will add the usage of async API's.

ramakrishnapallala commented:
@pizi-nordic I have addressed your comments and updated the PR. please have a look at it.

ramakrishnapallala commented:
> @pizi-nordic I have addressed your comments and updated the PR. please have a look at it.

@pizi-nordic can you look at the updated PR?

pizi-nordic commented:
@ramakrishnapallala: One moment. Please give me 1h :)

ramakrishnapallala commented:
> @ramakrishnapallala: One moment. Please give me 1h :)

sure,  thanks :)

pizi-nordic commented on samples/subsys/power/device_pm/src/dummy_driver.c:
```diff
@@ -0,0 +1,171 @@
+			printk("Dummy device resumed\n");
+			ret = 0;
+		} else {
+			printk("Dummy device Not resumed\n");
```
IMHO we have to wait here in a loop, as device might go through SUSPENDED state if put() was executed just before requesting get().

ramakrishnapallala replied:
which block should go in loop?

pizi-nordic replied:
We might get device in proper state in 2nd attempt. So whole waiting for event should be in loop in my opinion.

pizi-nordic commented on include/device.h:
```diff
@@ -208,6 +216,34 @@ extern "C" {
+	u32_t pm_status;
+	struct k_delayed_work dwork;
+	u32_t timeout;
+	struct k_poll_signal *signal;
```
Could we embed signal here? If we do so, driver will not need to implements its own.

pizi-nordic commented on samples/subsys/power/device_pm/src/dummy_driver.c:
```diff
@@ -0,0 +1,171 @@
+		printk("parent not found\n");
+	}
+
+	device_power_state = DEVICE_PM_ACTIVE_STATE;
```
IMHO this part of initialization might be part of common code.

ramakrishnapallala replied:
this can be device specific. For most cases, devices will be active but it cab be suspended as well.

pizi-nordic replied:
Then we will have to update drivers. Moreover the behaviour of drivers will change depending on the  device PM management setting (without: device active after init, with: suspended or active after init).

pizi-nordic commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,301 @@
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	dev->pm->enable = true;
+
+	/* During the driver init, device can set the
```
IMHO we have to have system-wide policy here.

ramakrishnapallala replied:
means? what do you suggest?

pizi-nordic replied:
IMHO all drivers should be obligated to leave the device in the active state after init. Per-driver policy will create a mess.

ramakrishnapallala replied:
The devices might be active but we need to call the pm_enable to init the device idle info. So the driver will need to call pm_enable during init or during config.

pizi-nordic replied:
This means, that every device driver will have to be updated.

pizi-nordic replied:
What about _sys_device_do_config_level() ?

ramakrishnapallala replied:
yeah, we can add in the for loop after device init is called.

pizi-nordic commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,301 @@
+}
+
+void device_pm_disable(struct device *dev)
+{
```
IMHO here we should wake-up the device regardless of the usage count.

ramakrishnapallala replied:
Ok got it.

pizi-nordic commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,301 @@
+	int ret = DEVICE_PM_RET_SUCCESS;
+
+	/* Get current device power state */
+	ret = device_get_power_state(dev, &current_state);
```
What this API return when device is changing power state now? (note, that power state change is not atomic operation).

ramakrishnapallala replied:
i think we need to make this atomic or control the flow with lock.

pizi-nordic commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,301 @@
+	}
+
+	if (dev->pm->usage > 0) {
+		ret = device_pm_state_handler(dev,
```
If the call is synchronous, it block all asynchronous API due to `dev->pm->lock`.

ramakrishnapallala replied:
Ok will see if i can improve the logic.

pizi-nordic commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,301 @@
+	 * dedicated lock to control power state transitins
+	 * between Sync and Async calls.
+	 */
+	k_sem_take(&dev->pm->pm_lock, K_FOREVER);
```
Here you can wait for `pm_lock` (good), holding `lock` (bad).

ramakrishnapallala replied:
if we don't hold the lock then the async worker will also try to change the state.

pizi-nordic replied:
I think that we have to decouple requests from the entity which do the change. All requests will be asynchronous (sync API will just wait for proper event after requesting change). Unfortunately this will need dedicated thread (we already have one for workq) unless we change device_set_power_state API.

ramakrishnapallala replied:
yes, I think we will end up there :)

pizi-nordic replied:
I personally would recommend making device_set_power_state  asynchronous. Some devices might  not have memory for additional thread stack.

ramakrishnapallala replied:
How do I make it asynchronous? queue the work on the same pm thread? or have a different thread?

pizi-nordic replied:
Are you asking the device_set_power_state()? I just would add a callback which should be called when the device will change the state. If another device_set_power_state() will be called during transition, the call will return -EINPROGRESS. If device is busy, and cannot change state at the moment, it could return -EBUSY.

ramakrishnapallala replied:
Sorry, I didn't get,  are you suggesting to change the  device_set_power_state() implementation? do we need to add the callback in device_pm struct?

pizi-nordic replied:
I would start from changing the  device_set_power_state() - to avoid dedicated thread in device PM management.

ramakrishnapallala replied:
Ok, will work in that direction.

ramakrishnapallala replied:
> Ok, will work in that direction.
> 
> @pizi-nordic Even if we want to make the device_set_power_state() asynchronous we would still need to queue the work either on system thread or pm thread or a new thread. This can be done from device pm code itself right. why do we need to change the device_set_power_state() API?

@pizi-nordic I have addressed all your comments except this one. I will spend a bit more time on this and get back to you by tomorrow.

pizi-nordic replied:
If you make the device_set_power_state() asynchronous you will be able to avoid dedicated thread. All logic might be triggered by either device_get()/device_put() or device_set_power_state() callback.

ramakrishnapallala replied:
@pizi-nordic I have rewritten some of the device_pm logic to decouple async request queuing and device set power state. Please have a look.

pizi-nordic reviewed (commented):
I apologize for the delay.
- Locking still could cause significant delays in async interface.
- Relying on device_get_power_state() is problematic, as we do not know what to return if device changes state. IMHO we have to track the device state externally and remove this API call.
- What about dependencies between devices?

ramakrishnapallala commented:
@pizi-nordic I have updated the PR, can you have a look. 
We need to close this by Zephyr feature freeze date (Jan 31).

ramakrishnapallala commented:
@pizi-nordic Can you please review the updated PR. Thanks.

pizi-nordic commented:
@ramakrishnapallala: I will try to go trough your PR today, however I am not sure if this will be possible. Since tomorrow I am out of the office, the next attempt will be in Monday.

ramakrishnapallala commented:
> @ramakrishnapallala: I will try to go trough your PR today, however I am not sure if this will be possible. Since tomorrow I am out of the office, the next attempt will be in Monday.

Sure, see what you can do. Thanks.

pdunaj commented on subsys/power/Kconfig:
```diff
@@ -17,9 +17,26 @@ config PM_CONTROL_OS_DEBUG
+config DEVPM_WORKQUEUE_STACK_SIZE
+	int "Device PM workqueue stack size"
+	depends on DEVICE_IDLE_PM
+	default 1024
```
1024B of RAM plus maybe a few hundred more per driver... just to count users... and I was looking how to free some memory in Zephyr just a week ago.
Thankfully this can be disabled and I hope this will not become mandatory.

ramakrishnapallala replied:
This is purely a config based option. The option can be enabled only if the application wants OS based PM and device idle management.

pdunaj replied:
True, but this would also require changes in the drivers. If drivers will implement it then application would have to use it as it will become a dependency.

pdunaj reviewed (changes requested):
If I understand correctly this is a common code to handle a driver power state. I had a look on the code and have following comments.

- The code significantly increases RAM footprint,
- I think the proposed code will not add a much value as driver itself will still need to track its state,
- For what the code does it is really complicated,
- I agree that bus drivers (such as I2C) need to count number of their users but this should be hierarchy based (i.e. driver should extend a dedicated bus API); not every driver needs this expensive logic.

Besides that:
- API names are misleading (e.g. device_pm_get/device_pm_put),
- Documentation is unclear,
- Text has poor quality (spelling issues).

I suggest agreeing first if we really want to bring this PR at all before doing any greater changes. I suggest checking with driver code authors if they will get any benefit from such common code instead of handling the code themselves.

From application point of view I don't see much benefit as I know I can disable i2c only when all its users are down. Personally I am not willing to pay over 1k of RAM for counting the driver usage. As for asynchronous suspending... this could be useful in some cases but our application needs to handle it anyway so I will have not benefit from that either.

ramakrishnapallala commented:
@pizi-nordic @pdunaj The child devices can do get/put operations from child's device suspend/resume calls. Can you elaborate more hierarchy support?

ramakrishnapallala commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,239 @@
+			goto req_out;
+		}
+	} else {
+		if (atomic_dec(&dev->pm->usage) > 1) {
```
In OS based PM and Device PM we are only dealing with ACTIVE and SUSPEND as of now. we need to unify all the states and have a common solution.

pdunaj replied:
if so then please assert that no other state was passed to avoid errors

ramakrishnapallala commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,239 @@
+		k_sem_take(&dev->pm->lock, K_FOREVER);
+		if (dev->pm->async_req_pending) {
+			request = dev->pm->async_request;
+			dev->pm->async_req_pending = false;
```
if there are back to ack call to these API's we need to know if any pending request is there while this work is being executed.

pdunaj replied:
you don't need a loop however as every time you set pending to true you are submitting the work

ramakrishnapallala commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,239 @@
+	/* Certain devices may take lot of time to switch
+	 * between power states. So to avoid blocking of
+	 * Async API's for longer periods we are using a
+	 * dedicated lock to control power state transitins
```
checkpatch was catching these typo's, somehow it is missed.

ramakrishnapallala commented on subsys/power/device_pm.c:
```diff
@@ -0,0 +1,239 @@
+	/* Bring up the device before disabling the Idle PM */
+	dev->pm->async_request = DEVICE_PM_ACTIVE_STATE;
+	dev->pm->async_req_pending = true;
+	k_delayed_work_submit_to_queue(&devpm_workq, &dev->pm->dwork, 0);
```
The worker should be executed from the thread defined in this file and also cancel/submit requests are taken care by checking the async request pending field.

pdunaj replied:
why not using system workqueue? if you need to sleep on poll you can poll with no wait and if nor ready resubmit work with delay (should save ram)
cancel is not needed if you resubmit btw

pdunaj replied:
some time ago I have created an issue with proposal that work could be submitted when polled object is ready (https://github.com/zephyrproject-rtos/zephyr/issues/10748)

pizi-nordic replied:
> why not using system workqueue?

The device_set_power_state() is synchronous and it could take 1s to change power state.
If we will use system workqueue, we will block system while executing state change.

IMHO we need asynchronous interface to device_set_power_state as a prerequisite to this PR.

pdunaj replied:
@pizi-nordic I agree. It seems that this PR is trying to address the limitation of the model in the drivers. First we should fix the API there otherwise the solution will never be right.

ramakrishnapallala replied:
> > why not using system workqueue?
> 
> The device_set_power_state() is synchronous and it could take 1s to change power state.
> If we will use system workqueue, we will block system while executing state change.
> 
> IMHO we need asynchronous interface to device_set_power_state as a prerequisite to this PR.

@pizi-nordic how can we achieve that? do we need to introduce new API's? what synchronous device_st_power_state requests? Ideally, we would need both right?

ramakrishnapallala commented:
@pizi-nordic @pdunaj can we take this PR as base implementation and take the hierarchy support as incremental change? The current implementation has support handling parent/child dependencies.

pdunaj commented:
One more note. This code needs to be used within the drivers otherwise it will be dead. If drivers will start to use it, then application would have to enable this feature too. For this reason we cannot say that this feature will be optional only.

As @pizi-nordic mentioned above we need to address problem of synchronous power mode switch inside the drivers. Hopefully then we will be able to create implementation that would be fast, clean and cheap.

ramakrishnapallala commented:
> One more note. This code needs to be used within the drivers otherwise it will be dead. If drivers will start to use it, then application would have to enable this feature too. For this reason we cannot say that this feature will be optional only.
> 
> As @pizi-nordic mentioned above we need to address problem of synchronous power mode switch inside the drivers. Hopefully then we will be able to create implementation that would be fast, clean and cheap.

@pdunaj even if we add the code to drivers, the application can still enable or disable this feature based on the need.

I will work with @pizi-nordic  to have a feasible solution.

Thanks.

nashif commented on samples/subsys/power/device_pm/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+cmake_minimum_required(VERSION 3.8.2)
```
change to 3.13.1

ramakrishnapallala replied:
Ok.

ramakrishnapallala commented:
@pizi-nordic @pdunaj @dbkinder @nashif I have updated the PR and added documentation.

ramakrishnapallala commented:
@dbkinder Can you check the rst file once, I am not sure why the documentation check is failing.

Commit:
power: Add device idle power management support

Add framework for device Idle Power Management(IPM)
for suspending devices based on device idle. This will
help in saving power even while system(CPU) is active.

This framework uses API's like device_set_power_state
and device_get_power_state to get/set the present
device power state. It relies on device_busy_check API
to know whether the device is idle or busy before
taking suspend or resume action.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

Commit:
samples: power: Add test for device Idle PM

Added test for Device Idle Power Management to invoke
device_pm_get/put API's.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

dbkinder commented on doc/reference/power_management/device_idle_pm.rst:
```diff
@@ -0,0 +1,105 @@
+
+
+The Device Idle Power Management framework is a Active Power
+Management mechanism which reduces the overall system Power consumtion
```
```suggestion
Management mechanism which reduces the overall system Power consumption
```

dbkinder commented on doc/reference/power_management/device_idle_pm.rst:
```diff
@@ -0,0 +1,105 @@
+Device Idle Power Management APIs
+=================================
+
+.. doxygengroup:: device_idle_power_management_api
```
The doc build is failing because there's no doxygen comment (@defgroup) defining the group ``device_idle_power_management_api`` that you're referencing here.  (There is one for ``device_power_managment_api``.)

ramakrishnapallala replied:
Thanks. Will fix it.

ramakrishnapallala replied:
@dbkinder I changed the API group but still failing.

nashif commented:
if i am not mistaken, this is superseded by #13382
